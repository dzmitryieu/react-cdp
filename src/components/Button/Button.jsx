import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.scss';

const Button = ({
  id,
  type,
  text,
  onClick,
  isMain,
  className,
}) => (
  <button id={id} className={`${styles.button} ${isMain ? styles.big : ''} ${styles[type]} ${className}`} type="button" onClick={onClick}>
    {text}
  </button>
);

Button.propTypes = {
  type: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  isMain: PropTypes.bool,
  className: PropTypes.string,
  id: PropTypes.string,
};
Button.defaultProps = {
  onClick: null,
  isMain: null,
  className: null,
  id: null,
};
export default Button;
