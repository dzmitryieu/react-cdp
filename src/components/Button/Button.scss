@import "../../utils/constants.scss";

.button {
  height: 30px;
  width: 100px;
  color: white;
  text-align: center;
  font-size: 14px;
  font-family: $helvetica;
  line-height: 22px;
  cursor: pointer;
  transition: all 0.3s;
  margin: 10px;
  outline: none;
  text-transform: uppercase;
  border: none;

  &:hover {
    cursor: pointer;
  }
}

.big {
  width: 150px;
  height: 40px;
  margin: 0;
}

.red {
  background-color: $light-red;

  &:hover {
    background-color: lighten($light-red, 20%);
  }
}

.black {
  background-color: $dark-gray;

  &:hover {
    color: darken(white, 20%);
  }
}
