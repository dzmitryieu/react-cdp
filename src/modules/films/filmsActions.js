import axios from 'axios';
import { cleanSearchField } from '../searchFilms/searchFilmsActions';

export const FETCH_FILMS_REQUEST = 'FETCH_FILMS_REQUEST';
export const FETCH_FILMS_SUCCESS = 'FETCH_FILMS_SUCCESS';
export const FETCH_FILMS_ERROR = 'FETCH_FILMS_ERROR';
export const SORT_FILMS = 'SORT_FILMS';

export const urlFilms = 'https://reactjs-cdp.herokuapp.com/movies';

export const receiveFilmsSuccess = payload => ({
  type: FETCH_FILMS_SUCCESS,
  payload,
});

export const receiveFilmsRequest = () => ({
  type: FETCH_FILMS_REQUEST,
});

export const receiveFilmsError = payload => ({
  type: FETCH_FILMS_ERROR,
  payload,
});

export const sortFilms = payload => ({
  type: SORT_FILMS,
  payload,
});

export function fetchFilms(url) {
  return (dispatch) => {
    dispatch(receiveFilmsRequest());
    return axios.get(url)
      .then((response) => {
        dispatch(receiveFilmsSuccess({ films: response.data.data, url }));
      })
      .catch(error => dispatch(receiveFilmsError(error.message)));
  };
}

export function fetchAllFilms() {
  return (dispatch) => {
    dispatch(cleanSearchField());
    return dispatch(fetchFilms(urlFilms));
  };
}

export function fetchFilmsBySearch(word, searchBy) {
  return (dispatch) => {
    dispatch(fetchFilms(`${urlFilms}?search=${word}&searchBy=${searchBy}`));
  };
}
