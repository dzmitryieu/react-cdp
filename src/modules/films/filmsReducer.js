import {
  FETCH_FILMS_REQUEST,
  FETCH_FILMS_SUCCESS,
  FETCH_FILMS_ERROR,
  SORT_FILMS,
} from './filmsActions';
import initialState from './index';
import sort from '../../utils/sort';

const filmsReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_FILMS_SUCCESS:
      return {
        ...state,
        isFetchingFilms: false,
        isFetchedFilms: true,
        allFilms: action.payload.films,
        url: action.payload.url,
        errorFilms: '',
      };
    case FETCH_FILMS_REQUEST:
      return {
        ...state,
        isFetchingFilms: true,
        isFetchedFilms: false,
      };
    case FETCH_FILMS_ERROR:
      return {
        ...state,
        errorFilms: action.payload,
        isFetchingFilms: false,
        isFetchedFilms: false,
        allFilms: [],
      };
    case SORT_FILMS:
      return {
        ...state,
        allFilms: [].concat(sort(state.allFilms, action.payload.direction, action.payload.key)),
      };
    default:
      return {
        ...state,
      };
  }
};

export default filmsReducer;
