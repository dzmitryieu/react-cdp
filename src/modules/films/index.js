const initialState = {
  allFilms: [],
  errorFilms: {},
  isFetchingFilms: false,
  isFetchedFilms: false,
  url: '',
  mainFilm: {},
};

export default initialState;
