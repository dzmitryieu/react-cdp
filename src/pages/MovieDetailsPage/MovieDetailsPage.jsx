import React from 'react';
import PropTypes from 'prop-types';
import queryString from 'query-string';
import styles from './MovieDetailsPage.scss';
import Header from '../../components/Header';
import MovieDetails from '../../components/MovieDetails';
import Button from '../../components/Button';
import '../../assets/images/cover-image.jpg';

class MovieDetailsPage extends React.Component {
  componentDidMount() {
    this.loadMainFilm();
  }

  componentDidUpdate(prevProps) {
    const { match } = this.props;
    if (prevProps.match !== match) {
      this.loadMainFilm();
    }
  }

  onClickHandler = () => {
    const {
      history,
      location,
    } = this.props;
    const parsed = queryString.parse(location.search);
    const searchWord = parsed.q;
    const searchBy = parsed.t;
    if (searchWord) {
      history.push(`/search?q=${searchWord}&t=${searchBy}`);
    } else {
      history.push('/');
    }
  };

  loadMainFilm = () => {
    const {
      match,
      location,
      history,
      fetchFilmByRoute,
      makeMainFilm,
      allFilms,
    } = this.props;
    const filmID = Number(match.params.id);
    const mainFilm = allFilms.find(film => film.id === filmID);
    if (mainFilm) {
      makeMainFilm(mainFilm);
    } else {
      fetchFilmByRoute(location, match, history);
    }
  }

  render() {
    const { mainFilm } = this.props;
    const isMainFilm = Object.keys(mainFilm).length !== 0;
    return (
      <div id="MovieDetailsPage" className={styles.wrapper}>
        <section className={styles.section}>
          <Header />
          <Button
            id="searchDetailsButton"
            type="white"
            text="SEARCH"
            isMain
            className={styles.searchButton}
            onClick={this.onClickHandler}
          />
          {isMainFilm && <MovieDetails {...mainFilm} />}
          {!isMainFilm && <h3 className={styles.title}>There is no such movie...</h3>}
        </section>
      </div>
    );
  }
}

MovieDetailsPage.propTypes = {
  allFilms: PropTypes.arrayOf(PropTypes.any),
  mainFilm: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    releaseDate: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    posterPath: PropTypes.string,
  }),
  match: PropTypes.shape({
    params: PropTypes.object,
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
  }).isRequired,
  history: PropTypes.objectOf(PropTypes.any).isRequired,
  fetchFilmByRoute: PropTypes.func.isRequired,
  makeMainFilm: PropTypes.func.isRequired,
};

MovieDetailsPage.defaultProps = {
  allFilms: {},
  mainFilm: PropTypes.shape({
    id: null,
    name: null,
    releaseDate: null,
    genres: null,
    posterPath: null,
  }),
};

export default MovieDetailsPage;
