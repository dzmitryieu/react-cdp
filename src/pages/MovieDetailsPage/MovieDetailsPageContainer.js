import { connect } from 'react-redux';
import MovieDetailsPage from './MovieDetailsPage';
import { showSearchPage } from '../../modules/films/filmsActions';
import { getMainFilm, isShowSearchPage } from '../../modules/films/filmsSelectors';


const mapStateToProps = state => ({
  mainFilm: getMainFilm(state),
  isShowSearchPage: isShowSearchPage(state),
});

const mapDispatchToProps = {
  showSearchPage,
};

export default connect(mapStateToProps, mapDispatchToProps)(MovieDetailsPage);
