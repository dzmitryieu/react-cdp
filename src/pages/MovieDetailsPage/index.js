import MovieDetailsPage from './MovieDetailsPageContainer';

export default MovieDetailsPage;
