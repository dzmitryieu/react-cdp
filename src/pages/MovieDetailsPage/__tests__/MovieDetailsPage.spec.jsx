import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Testrenderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MovieDetailsPage from '../MovieDetailsPage';
import MovieDetailsPageContainer from '../index';
import Button from '../../../components/Button';
import films from '../../../modules/mocks/getFilmsMock.json';

const renderer = new ShallowRenderer();
Enzyme.configure({ adapter: new Adapter() });
const film = {
  id: 13456,
  title: 'Captain Marvel',
  poster_path: '../../assets/images/poster.jpg',
  genres: ['Drama', 'Comedy'],
  release_date: '1984-12-03',
  runtime: 150,
  tagline: 'Super movie',
  overview: 'Super movie sdfsfs',
};

const allFilms = films.data;

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  film: {
    mainFilm: film,
  },
  films: { allFilms },
});

const match = {
  params: {
    id: 447365,
  },
};

const match1 = {
  params: {
    id: 123,
  },
};

const history = {
  push: jest.fn(),
};
const location = {};
const location2 = {
  search: 'q=rambo&t=title',
};

const fetchFilmByRoute = jest.fn();
const makeMainFilm = jest.fn();

const emptyFilm = {};

describe('MovieDetailsPage Snapshot', () => {
  test('renders', () => {
    const component = renderer.render(
      <MovieDetailsPage
        match={match}
        makeMainFilm={makeMainFilm}
        allFilms={allFilms}
        fetchFilmByRoute={fetchFilmByRoute}
        history={history}
        mainFilm={film}
        location={location}
      />,
    );
    expect(component).toMatchSnapshot();
  });
  test('with store render', () => {
    const component = Testrenderer.create(
      <Provider store={store}>
        <Router>
          <MovieDetailsPageContainer
            match={match}
            location={location}
            history={history}
          />
        </Router>
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
});

describe('MovieDetailsPage handler should work', () => {
  it('click to the search button should call clickHandler and show nothing if mainFilm is empty', () => {
    const wrapper = shallow(
      <MovieDetailsPage
        match={match}
        makeMainFilm={makeMainFilm}
        allFilms={allFilms}
        fetchFilmByRoute={fetchFilmByRoute}
        history={history}
        mainFilm={emptyFilm}
        location={location}
      />,
    );
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'onClickHandler');
    wrapper.instance().forceUpdate();
    wrapper.find(Button).prop('onClick')();
    expect(spy).toHaveBeenCalled();
  });
  it('click to the search button should call clickHandler film not found', () => {
    const wrapper = shallow(
      <MovieDetailsPage
        match={match1}
        makeMainFilm={makeMainFilm}
        allFilms={allFilms}
        fetchFilmByRoute={fetchFilmByRoute}
        history={history}
        location={location2}
        mainFilm={emptyFilm}
      />,
    );
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'onClickHandler');
    wrapper.instance().forceUpdate();
    wrapper.find(Button).prop('onClick')();
    expect(spy).toHaveBeenCalled();
  });
  it('should make an extra request if URL changed', () => {
    const wrapper = shallow(
      <MovieDetailsPage
        match={match}
        allFilms={allFilms}
        makeMainFilm={makeMainFilm}
        fetchFilmByRoute={fetchFilmByRoute}
        history={history}
        location={location}
        mainFilm={film}
      />,
    );
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'loadMainFilm');
    wrapper.setProps({ match: { params: { id: 777 } } });
    expect(spy).toBeCalled();
  });
  it('should make an extra request if URL not changed', () => {
    const wrapper = shallow(
      <MovieDetailsPage
        match={match}
        allFilms={allFilms}
        makeMainFilm={makeMainFilm}
        fetchFilmByRoute={fetchFilmByRoute}
        history={history}
        location={location}
        mainFilm={film}
      />,
    );
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'loadMainFilm');
    wrapper.setProps(match);
    expect(spy).not.toBeCalled();
  });
});
