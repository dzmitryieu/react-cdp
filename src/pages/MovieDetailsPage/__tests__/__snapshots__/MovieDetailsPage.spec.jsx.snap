// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsPage Snapshot renders 1`] = `
<div
  className="wrapper"
  id="MovieDetailsPage"
>
  <section
    className="section"
  >
    <Connect(Header) />
    <Button
      className="searchButton"
      dataAttr={null}
      id="searchDetailsButton"
      isMain={true}
      onClick={[Function]}
      text="SEARCH"
      type="white"
    />
    <MovieDetails
      genres={
        Array [
          "Drama",
          "Comedy",
        ]
      }
      id={13456}
      overview="Super movie sdfsfs"
      poster_path="../../assets/images/poster.jpg"
      release_date="1984-12-03"
      runtime={150}
      tagline="Super movie"
      title="Captain Marvel"
    />
  </section>
</div>
`;

exports[`MovieDetailsPage Snapshot with store render 1`] = `
<div
  className="wrapper"
  id="MovieDetailsPage"
>
  <section
    className="section"
  >
    <header
      className="header"
    >
      <a
        className="link"
        href="/"
        onClick={[Function]}
      >
        <p
          className="logo"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="presentation"
        >
          netflixroulette
        </p>
      </a>
    </header>
    <button
      className="button big white searchButton"
      data-attr={null}
      id="searchDetailsButton"
      onClick={[Function]}
      type="button"
    >
      SEARCH
    </button>
    <div
      className="container"
    >
      <div
        className="wrapper"
      >
        <div
          className="poster"
          style={
            Object {
              "backgroundImage": "url(../../assets/images/poster.jpg)",
            }
          }
        />
        <div
          className="description"
        >
          <h2
            className="title"
          >
            Captain Marvel
          </h2>
          <h4
            className="details"
          >
            Super movie
          </h4>
          <div
            className="statistics"
          >
            <p
              className="year"
            >
              1984
            </p>
            <p
              className="duration"
            >
              150 min
            </p>
          </div>
          <p
            className="shortStory"
          >
            Super movie sdfsfs
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;
